lib/client/mfa/prompt.go

```diff
@@ -119,182 +144,72 @@ func (p *Prompt) Run(ctx context.Context, chal *proto.MFAAuthenticateChallenge)
 	}
 
-	// Use variables below to cancel OTP reads and make sure the goroutine exited.
-	otpWait := &sync.WaitGroup{}
-	otpCtx, otpCancel := context.WithCancel(ctx)
-	defer otpCancel()
-
-	// Fire TOTP goroutine.
-	if hasTOTP {
-		otpWait.Add(1)
-		wg.Add(1)
-		go func() {
-			defer otpWait.Done()
-			defer wg.Done()
-			const kind = "TOTP"
-
-			// Let Webauthn take the prompt, it knows better if it's necessary.
-			var msg string
-			if !quiet && !hasWebauthn {
-				msg = fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix)
-			}
+	// No challenge present, return an empty response.
+	return &proto.MFAAuthenticateResponse{}, nil
+}
 
-			otp, err := prompt.Password(otpCtx, writer, prompt.Stdin(), msg)
-			if err != nil {
-				respC <- response{kind: kind, err: err}
-				return
-			}
-			respC <- response{
-				kind: kind,
-				resp: &proto.MFAAuthenticateResponse{
-					Response: &proto.MFAAuthenticateResponse_TOTP{
-						TOTP: &proto.TOTPResponse{Code: otp},
-					},
-				},
-			}
-		}()
+func (c PromptConfig) promptDevicePrefix() string {
+	if c.DeviceType != "" {
+		return fmt.Sprintf("*%s* ", c.DeviceType)
 	}
+	return ""
+}
 
-	// Fire Webauthn goroutine.
-	if hasWebauthn {
-		origin := p.ProxyAddress
-		if !strings.HasPrefix(origin, "https://") {
-			origin = "https://" + origin
-		}
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			log.Debugf("WebAuthn: prompting devices with origin %q", origin)
-
-			prompt := wancli.NewDefaultPrompt(ctx, writer)
-			prompt.SecondTouchMessage = fmt.Sprintf("Tap your %ssecurity key to complete login", promptDevicePrefix)
-			switch {
-			case quiet:
-				// Do not prompt.
-				prompt.FirstTouchMessage = ""
-				prompt.SecondTouchMessage = ""
-			case hasTOTP: // Webauthn + OTP
-				prompt.FirstTouchMessage = fmt.Sprintf("Tap any %ssecurity key or enter a code from a %sOTP device", promptDevicePrefix, promptDevicePrefix)
-
-				// Customize Windows prompt directly.
-				// Note that the platform popup is a modal and will only go away if
-				// canceled.
-				webauthnwin.PromptPlatformMessage = "Follow the OS dialogs for platform authentication, or enter an OTP code here:"
-				defer webauthnwin.ResetPromptPlatformMessage()
-
-			default: // Webauthn only
-				prompt.FirstTouchMessage = fmt.Sprintf("Tap any %ssecurity key", promptDevicePrefix)
-			}
-			mfaPrompt := &mfaPrompt{LoginPrompt: prompt, otpCancelAndWait: func() {
-				otpCancel()
-				otpWait.Wait()
-			}}
-
-			resp, _, err := p.WebauthnLogin(ctx, origin, wantypes.CredentialAssertionFromProto(chal.WebauthnChallenge), mfaPrompt, &wancli.LoginOpts{
-				AuthenticatorAttachment: p.AuthenticatorAttachment,
-			})
-			respC <- response{kind: "WEBAUTHN", resp: resp, err: err}
-		}()
+// WebauthnLogin performs client-side Webauthn login from prompt settings.
+func (c PromptConfig) WebauthnLogin(ctx context.Context, chal *proto.MFAAuthenticateChallenge, loginPrompt wancli.LoginPrompt) (*proto.MFAAuthenticateResponse, error) {
```

Comment:
Could we push this logic to Prompt instead? It seems off to me having this belong to the config.